core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -345,23 +347,13 @@ private void checkAndAdvanceBlockInStream() throws IOException {
         try {
           WorkerNetAddress address = mLocationPolicy.getWorkerForNextBlock(
               mContext.getAlluxioBlockStore().getWorkerInfoList(), getBlockSizeAllocation(mPos));
-          // Don't cache the block to somewhere that already has it.
           // TODO(andrew): Filter the workers provided to the location policy to not include
           // workers which already contain the block. See ALLUXIO-1816.
-          if (mCurrentBlockInStream instanceof RemoteBlockInStream) {
-            WorkerNetAddress readAddress =
-                ((RemoteBlockInStream) mCurrentBlockInStream).getWorkerNetAddress();
-            // Try to avoid an RPC.
-            if (readAddress.equals(address)) {
-              mShouldCacheCurrentBlock = false;
-            } else {
-              BlockInfo blockInfo = mContext.getAlluxioBlockStore().getInfo(currentBlockId);
-              for (BlockLocation location : blockInfo.getLocations()) {
-                if (address.equals(location.getWorkerAddress())) {
-                  mShouldCacheCurrentBlock = false;
-                }
-              }
-            }
+          if (mCurrentBlockInStream instanceof RemoteBlockInStream
+              && !address.getHost().equals(mLocalHostName)) {
+            // If the block is already on some remote worker, then we will try to cache it only on
+            // the local host if possible.
+            mShouldCacheCurrentBlock = false;
```

Comment:
@yupeng9 Yes. As the jira ticket says, currently Alluxio is not topology-aware. It seems to make no sense if we cache a block at a remote worker when reading it from another remote worker since this behavior will consume memory and evict other blocks.